pandas/io/parsers.py

```diff
@@ -786,17 +786,25 @@ def read(self, nrows=None):
             return ret
 
         # May alter columns / col_dict
-        index, columns, col_dict = self._create_index(ret)
+        index, columns, mappings, col_dict = self._create_index(ret)
```

Comment:
`mappings`: `names` --> unique labels
`col_dict` is keyed using the unique labels.  This way I can pass in `col_dict` into the `DataFrame` constructor.  As you can see then at the end, I replace the unique labels with the `names` using `mappings`.